crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -100,55 +244,341 @@ fn site_packages_dir_from_sys_prefix(
+                f.write_str("the value for the `home` key does not point to a directory on disk")
+            }
+        }
+    }
+}
+
 #[derive(Debug, thiserror::Error)]
 pub enum SitePackagesDiscoveryError {
+    #[error("Invalid --venv-path argument: {0} does not point to a directory on disk")]
+    VenvDirIsNotADirectory(SystemPathBuf),
+    #[error("--venv-path points to a broken venv with no pyvenv.cfg file")]
+    NoPyvenvCfgFile(#[source] io::Error),
+    #[error("Failed to parse the pyvenv.cfg file at {0} because {1}")]
+    PyvenvCfgParseError(SystemPathBuf, PyvenvCfgParseErrorKind),
     #[error("Failed to search the virtual environment directory for `site-packages`")]
-    CouldNotReadLibDirectory(#[from] io::Error),
+    CouldNotReadLibDirectory(#[source] io::Error),
     #[error("Could not find the `site-packages` directory in the virtual environment")]
     NoSitePackagesDirFound,
 }
 
-/// Given a validated, canonicalized path to a virtual environment,
-/// return a list of `site-packages` directories that are available from that environment.
-///
-/// See the documentation for `site_packages_dir_from_sys_prefix` for more details.
+/// Abstraction for a Python virtual environment.
 ///
-/// TODO: Currently we only ever return 1 path from this function:
-/// the `site-packages` directory that is actually inside the virtual environment.
-/// Some `site-packages` directories are able to also access system `site-packages` and
-/// user `site-packages`, however.
-pub fn site_packages_dirs_of_venv(
-    venv_path: &SystemPath,
-    system: &dyn System,
-) -> Result<Vec<SystemPathBuf>, SitePackagesDiscoveryError> {
-    Ok(vec![site_packages_dir_from_sys_prefix(venv_path, system)?])
+/// Most of this information is derived from the virtual environment's `pyvenv.cfg` file.
+/// The format of this file is not defined anyway, and exactly which keys are present
+/// depends on the tool that was used to create the virtual environment.
+#[derive(Debug)]
+pub struct VirtualEnvironment {
+    venv_path: SysPrefixPath,
+    base_executable_home_path: PythonHomePath,
+    include_system_site_packages: bool,
+
+    /// The minor version of the Python executable that was used to create this virtual environment.
+    ///
+    /// The Python version is encoded under different keys and in different formats
+    /// by different virtual-environment creation tools,
+    /// and the key is never read by the standard-library `site.py` module,
+    /// so it's possible that we might not be able to find this information
+    /// in an acceptable format under any of the keys we expect.
+    /// This field will be `None` if so.
+    minor_version: Option<PythonMinorVersion>,
+}
+
+impl VirtualEnvironment {
+    pub fn new(
+        path: impl Into<SystemPathBuf>,
+        system: &dyn System,
+    ) -> Result<Self, SitePackagesDiscoveryError> {
+        Self::new_impl(path.into(), system)
+    }
+
+    fn new_impl(
```

Comment:
What's the benefit of splitting this out? (Not a problem, just curious)

Reply:
https://github.com/astral-sh/ruff/pull/12759#discussion_r1709913740

Reply:
Aha, very interesting, I hadn't considered that!